src/core/cm/imagemanager/imagemanager.cpp

```diff
@@ -0,0 +1,949 @@
+    }
+
+    if (versionResult < 0) {
+        return ErrorEnum::eWrongState;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ImageManager::ValidateCachedVersionItem(const UpdateItemInfo& itemInfo, const Array<storage::ItemInfo>& items)
+{
+    LOG_DBG() << "Validate cached version item" << Log::Field("ID", itemInfo.mID)
+              << Log::Field("version", itemInfo.mVersion);
+
+    auto it = items.FindIf(
+        [&itemInfo](const storage::ItemInfo& item) { return item.mState == storage::ItemStateEnum::eCached; });
+
+    if (it == items.end()) {
+        return ErrorEnum::eNotFound;
+    }
+
+    auto [versionResult, err] = semver::CompareSemver(itemInfo.mVersion, it->mVersion);
+
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (versionResult == 0) {
+        if (auto setStateErr = SetState(*it, storage::ItemStateEnum::eActive); !setStateErr.IsNone()) {
+            return AOS_ERROR_WRAP(setStateErr);
+        }
+
+        return ErrorEnum::eAlreadyExist;
+    }
+
+    if (versionResult > 0) {
+        if (auto removeErr = Remove(*it); !removeErr.IsNone()) {
+            return AOS_ERROR_WRAP(removeErr);
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ImageManager::SetState(const storage::ItemInfo& item, storage::ItemState state)
+{
+    LOG_DBG() << "Set state" << Log::Field("ID", item.mID) << Log::Field("Version", item.mVersion)
+              << Log::Field("State", state);
+
+    if (auto err = mStorage->SetItemState(item.mID, item.mVersion, state); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (state == storage::ItemStateEnum::eCached) {
+        if (auto err = mSpaceAllocator->AddOutdatedItem(item.mID, item.mTotalSize, item.mTimestamp); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+    }
+
+    if (item.mState == storage::ItemStateEnum::eCached) {
+        if (auto err = mSpaceAllocator->RestoreOutdatedItem(item.mID); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ImageManager::Remove(const storage::ItemInfo& item)
+{
+    LOG_DBG() << "Removing item" << Log::Field("ID", item.mID) << Log::Field("Version", item.mVersion);
+
+    NotifyItemRemovedListeners(item.mID);
+
+    if (auto err = fs::RemoveAll(item.mPath); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (item.mState == storage::ItemStateEnum::eCached) {
+        if (auto err = mSpaceAllocator->RestoreOutdatedItem(item.mID); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+    }
+
+    mSpaceAllocator->FreeSpace(item.mTotalSize);
+
+    if (auto err = mStorage->RemoveItem(item.mID, item.mVersion); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    LOG_DBG() << "Item removed" << Log::Field("ID", item.mID) << Log::Field("Version", item.mVersion);
+
+    return ErrorEnum::eNone;
+}
+
+void ImageManager::NotifyItemRemovedListeners(const String& id)
+{
```

Comment:
Should be locked to guard mListeners

Reply:
this function called from Remove method which always called under mutex